File: spring-boot-jdbc/src/main/kotlin/hvalfangst/heroes/repository/HeroRepository.kt
package hvalfangst.heroes.repository

import hvalfangst.heroes.model.Hero
import hvalfangst.heroes.model.tables.HeroesTable
import org.jetbrains.exposed.sql.ResultRow
import org.jetbrains.exposed.sql.select
import org.jetbrains.exposed.sql.transactions.transaction
import org.springframework.stereotype.Repository

@Repository
class HeroRepository {

    fun getHeroById(id: Int): Hero? = transaction {
        HeroesTable.select { HeroesTable.id eq id }.singleOrNull()?.let { toHero(it) }
    }

    private fun toHero(row: ResultRow): Hero {
        return Hero(
            id = row[HeroesTable.id],
            name = row[HeroesTable.name],
            hitPoints = row[HeroesTable.hitPoints],
            dateOfBirth = row[HeroesTable.dateOfBirth],
            race = row[HeroesTable.race],
            characterClass = row[HeroesTable.characterClass],
            level = row[HeroesTable.level],
            alignment = row[HeroesTable.alignment],
            strength = row[HeroesTable.strength],
            dexterity = row[HeroesTable.dexterity],
            constitution = row[HeroesTable.constitution],
            intelligence = row[HeroesTable.intelligence],
            wisdom = row[HeroesTable.wisdom],
            charisma = row[HeroesTable.charisma],
            armorClass = row[HeroesTable.armorClass],
            speed = row[HeroesTable.speed],
            initiative = row[HeroesTable.initiative],
            proficiencyBonus = row[HeroesTable.proficiencyBonus],
            savingThrows = row[HeroesTable.savingThrows],
            skills = row[HeroesTable.skills],
            languages = row[HeroesTable.languages],
            features = row[HeroesTable.features],
            equipment = row[HeroesTable.equipment],
            background = row[HeroesTable.background],
            attack = row[HeroesTable.attack],
            damage = row[HeroesTable.damage]
        )
    }
}

File: go-fasthttp-pgx/src/repository/hero/hero_repository.go
package hero

import (
	"context"
	"errors"
	"github.com/jackc/pgx/v4"
	"github.com/jackc/pgx/v4/pgxpool"
	"hvalfangst/golang-fasthttp-pgx/src/model"
	"log"
)

func GetHeroById(ctx context.Context, conn *pgxpool.Conn, heroId int64) (*model.Heroes, error) {
	hero := &model.Heroes{}

	err := conn.QueryRow(
		ctx,
		"SELECT id, name, hit_points, date_of_birth, race, hero_class, level, alignment, "+
			"strength, dexterity, constitution, intelligence, wisdom, charisma, armor_class, "+
			"speed, initiative, proficiency_bonus, saving_throws, skills, languages, features, "+
			"equipment, background, attack, damage FROM heroes WHERE id = $1",
		heroId,
	).Scan(
		&hero.ID, &hero.Name, &hero.HitPoints, &hero.DateOfBirth, &hero.Race, &hero.HeroClass,
		&hero.Level, &hero.Alignment, &hero.Strength, &hero.Dexterity, &hero.Constitution,
		&hero.Intelligence, &hero.Wisdom, &hero.Charisma, &hero.ArmorClass, &hero.Speed,
		&hero.Initiative, &hero.ProficiencyBonus, &hero.SavingThrows, &hero.Skills,
		&hero.Languages, &hero.Features, &hero.Equipment, &hero.Background, &hero.Attack,
		&hero.Damage,
	)

	if errors.Is(err, pgx.ErrNoRows) {
		log.Printf("Hero with ID %d not found", heroId)
		return nil, nil // Return nil and no error for not found
	} else if err != nil {
		log.Printf("Error retrieving hero by ID: %v", err)
		return nil, err
	}

	// If no error occurred, return the hero.
	return hero, nil
}
